zetaclient/bitcoin_client.go

```diff
@@ -719,7 +735,55 @@ func (ob *BitcoinChainClient) observeOutTx() {
 	}
 }
 
-func (ob *BitcoinChainClient) BuildSubmittedTxMap() error {
+// Returns true if outTx passes basic checks
+func (ob *BitcoinChainClient) checkTssOutTxResult(hash *chainhash.Hash, res *btcjson.GetTransactionResult) error {
+	if res.Confirmations == 0 {
+		rawtx, err := ob.rpcClient.GetRawTransactionVerbose(hash) // for pending tx, we query the raw tx
+		if err != nil {
+			return errors.Wrapf(err, "checkTssOutTxResult: error GetRawTransactionVerbose %s", res.TxID)
+		}
+		if len(rawtx.Vin) == 0 {
+			return errors.Wrapf(err, "checkTssOutTxResult: invalid outTx with no vin %s", res.TxID)
+		}
+		if !ob.isTSSVin(rawtx.Vin) {
+			return errors.Wrapf(err, "checkTssOutTxResult: invalid outTx with non-TSS vin %s", res.TxID)
+		}
+	} else if res.Confirmations > 0 {
+		blkHash, err := chainhash.NewHashFromStr(res.BlockHash)
+		if err != nil {
+			return errors.Wrapf(err, "checkTssOutTxResult: error NewHashFromStr %s", res.BlockHash)
+		}
+		block, err := ob.rpcClient.GetBlockVerboseTx(blkHash) // for confirmed tx, we query the block
+		if err != nil {
+			return errors.Wrapf(err, "checkTssOutTxResult: error GetBlockVerboseTx %s", res.BlockHash)
+		}
+		if res.BlockIndex < 0 || res.BlockIndex >= int64(len(block.Tx)) {
+			return errors.Wrapf(err, "checkTssOutTxResult: invalid outTx with invalid block index, TxID %s, BlockIndex %d", res.TxID, res.BlockIndex)
+		}
+		tx := block.Tx[res.BlockIndex]
+		if !ob.isTSSVin(tx.Vin) {
+			return errors.Wrapf(err, "checkTssOutTxResult: invalid outTx with non-TSS vin %s", res.TxID)
+		}
+	}
+	return nil // ignore res.Confirmations < 0
+}
+
+// Returns true only if all inputs are TSS vins
+func (ob *BitcoinChainClient) isTSSVin(vins []btcjson.Vin) bool {
+	pubKeyTss := hex.EncodeToString(ob.Tss.PubKeyCompressedBytes())
+	for _, vin := range vins {
+		// The length of the Witness should be always 2 for P2WPKH SegWit inputs.
+		if len(vin.Witness) != 2 {
+			return false
+		}
+		if vin.Witness[1] != pubKeyTss {
+			return false
+		}
+	}
+	return true
+}
+
+func (ob *BitcoinChainClient) BuildIncludedTxMap() error {
```

Comment:
I would add documentation for each public method

Reply:
I agree. For this one, I'm trying to get rid of this unnecessary persistence.